### PR TITLE
fix(AIRMap): include map rotation in zoom computing & fix flyovers maps zoom constrains [take II]

### PR DESCRIPTION
**Pull request by jan-kozinski**
### Does any other open PR do the same thing?

No

### What issue is this PR fixing?

https://github.com/react-native-maps/react-native-maps/issues/4880

The zoom limit algorithm does not take map rotation into account and also it does not work for flyover maps. Flyover maps are not typed and not documented.

### How did you test this PR?

iPhone 14 Pro simulator, I logged the zoom values with NSLog. The new algorithm produced same zoom values as the old one did when map was not rotated. When the map was rotated, the old algorithm tended to be unpredictable.

### CODE REVIEW

**Comment by jan-kozinski** [comment]
@salah-ghanim this is take two of closed PR https://github.com/react-native-maps/react-native-maps/pull/4892

**Comment by salah-ghanim** [comment]
@jan-kozinski thanks a lot for re-submitting the pull request.

I added couple of comments that I think need to be addressed before merging the code. specifically I believe there is a way to implement max/min zoom based on having a setter function for both properties and calculating how much that would be in meters (based on screen dimensions).

let me know if you can take care of that or if I should try to do it myself.

**Comment by jan-kozinski** [comment]
@salah-ghanim Hi, I can't find the comments, perhaps you didn't submit the review yet? :D I'll do my best to address them, though I'm not sure how to compute the distance in meters based on screen dimensions.

**Comment by salah-ghanim** [comment]
@jan-kozinsk you should be able to see all comments in the "files changed" tab of the pull request.

**Comment by ahinkle** [comment]
@salah-ghanim I don't see your comments either. Would be happy to help with this PR :)

**Review by salah-ghanim — CHANGES_REQUESTED** [review]

**Comment by salah-ghanim** [comment]
@jan-kozinski @ahinkle strange that you guys don't see the feedback, I see it even in guest mode:
https://github.com/react-native-maps/react-native-maps/pull/4905#pullrequestreview-1774078137

anyhow I also included screenshots of the comments so that you see my feedback

Screenshot 2024-01-21 at 18 22 44
Screenshot 2024-01-21 at 18 22 21
Screenshot 2024-01-21 at 18 21 48

File: ios/AirMaps/AIRMap.m [hunk 1]
```diff
@@ -675,4 +679,90 @@ - (void)layoutSubviews {
     }
 }
 
+// based on https://medium.com/@dmytrobabych/getting-actual-rotation-and-zoom-level-for-mapkit-mkmapview-e7f03f430aa9
+- (double)getZoomLevel {
```
[the remark this answers is not in the document]
Comment (jan-kozinski) [reply ios/AirMaps/AIRMap.m:683]: well, this `getZoomLevel` code isn't for 3d maps actually. It is used for the regular maps as I found them to not take rotation into zoom calculations and thus the max / min zoom wasn't working as expected when rotation is twisted.

If by `keep updating max / min` you refer to lines 1165-1173 of AIRMapManager.m then you're right - we can set the camera range once, there's no need to include the check in the regionDidChange callback. I'll commit an improvement in a moment

**Comment by jan-kozinski** [comment]
Hi Salah,
thanks for resubmitting the comments, they are visible now. And
great review btw, I didn't know I was using a private API. I did my best to
address your comments so you can give it a second look when you have a
minute.
All the best

niedz., 21 sty 2024 o 18:24 Salah Ghanim ***@***.***>
napisał(a):

**Comment by salah-ghanim** [comment]
@jan-kozinski thank you so much for the new pull request, but I realised that:
1. we're still doing calculations each time mapRegion changes when we could avoid that by using MKMapView's constraints functionality.
2. we're adding "magic code" / opinionated that maps zoom level to visible meters that I personally can't confirm to match especially knowing that there are multiple device screen sizes out there so, seeing: 
`CLLocationDistance minDistance = pow(2, (26 - maxZoomLevel));`

doesn't set with me well.

but good news I see the way forward and it's much simpler that what we're trying to do.

1. we should expose new Properties for iOS zoom that exactly match iOS APIs: MKMapCameraZoomRange (min, max) in meters and make them available to JS / react-native users.

if such values are provided you can set the MapKit camera zoom range using those and avoid the use of the current zoom logic

2. disable the current zoom logic for satellite maps completely.
3. use Apple APIs for zoom whenever the min / max params are provided and stop trying to control the zoom level on each regionChange ... apple will ensure that the camera is limited since you provide the limitation.


by "current logic" I mean the current one in the library not the one in this pull request.


maybe to summarise again:
1. disable old logic for satellite map to avoid the bug we know about.
2. introduce new Props for iOS maps that match exactly MKMapView.cameraRange: https://developer.apple.com/documentation/mapkit/mkmapview/3114301-camerazoomrange?language=objc
3. if cameraRange is used, disable the current "reactive / onRegionChange" logic because iOS internals should take care of the limitation.

**Comment by salah-ghanim** [comment]
@jan-kozinski thanks for the new update some tiny feedback:
1. reactiveZoomConstraintsEnabled I would rename it to: legacyZoomConstraintslEnabled 
2. in the doc, please don't mention satellite as a reason why minCenterCoordinateDistance / maxCenterCoordinateDistance .. rather this is the new APIs and should be used, please mark old zoom values for deprecations.
please make sure to note that this is only relevant for iOS MapKit

once done I think we can finally merge :)

**Comment by jan-kozinski** [comment]
Done @salah-ghanim. I didn't use the `@deprecated` marker for old zoom levels properties in Mapview.tsx, as those properties don't have any alternative for google maps. Though I added a note that it is deprecated on Apple Maps

**Review by salah-ghanim — APPROVED** [review]

**Comment by ahinkle** [comment]
@jan-kozinski,

I'm still seeing this bug on the `hybridFlyover` map type. Any ideas there?

https://github.com/react-native-maps/react-native-maps/assets/17038330/3921b70e-1d6c-47dc-ba30-ef4d203d42a1

**Comment by jan-kozinski** [comment]
I'm not sure @ahinkle as the code runs well in the example app on current master. I experience the same bug after installing the newest version of react native maps in a clean repo though.

**Comment by jan-kozinski** [comment]
Hi @ahinkle I think you're using Expo. If so, you have to rebuild your project, not only update the library and start

**Pull request merged**